libs/backend/src/scan/cast_vote_records/validation.ts

```diff
@@ -0,0 +1,282 @@
+  Precinct,
+  SheetOf,
+} from '@votingworks/types';
+import { throwIllegalValue, Result, ok, err } from '@votingworks/basics';
+
+/**
+ * The back of a BMD ballot should be empty, which could be interpreted as
+ * either a blank or an unreadable page.
+ */
+const EMPTY_PAGE_TYPES: ReadonlyArray<PageInterpretation['type']> = [
+  'BlankPage',
+  'UnreadablePage',
+];
+
+/**
+ * Sheets with page interpretations of types outside this list can be excluded
+ * immediately because they are either uninterpreted or already invalid.
+ */
+const VALID_PAGE_TYPES: ReadonlyArray<PageInterpretation['type']> = [
+  ...EMPTY_PAGE_TYPES,
+  'InterpretedBmdPage',
+  'InterpretedHmpbPage',
+];
+
+/**
+ * Enumeration representing possible sheet validation errors
+ */
+export enum SheetValidationErrorType {
+  InvalidPageType = 'InvalidPageType',
+  InvalidFrontBackPageTypes = 'InvalidFrontBackPageTypes',
+  MismatchedBallotStyle = 'MismatchedBallotStyle',
+  MismatchedBallotType = 'MismatchedBallotType',
+  MismatchedElectionHash = 'MismatchedElectionHash',
+  MismatchedLocales = 'MismatchedLocales',
+  MismatchedPrecinct = 'MismatchedPrecinct',
+  NonConsecutivePages = 'NonConsecutivePages',
+}
+
+/**
+ * Type containing a validation error and information relevant to that error
+ */
+export type SheetValidationError =
+  | {
+      type: SheetValidationErrorType.NonConsecutivePages;
+      pageNumbers: SheetOf<HmpbBallotPageMetadata['pageNumber']>;
+    }
+  | {
+      type: SheetValidationErrorType.InvalidFrontBackPageTypes;
+      types: SheetOf<PageInterpretation['type']>;
+    }
+  | {
+      type: SheetValidationErrorType.MismatchedBallotStyle;
+      ballotStyleIds: SheetOf<BallotStyle['id']>;
+    }
+  | {
+      type: SheetValidationErrorType.MismatchedBallotType;
+      ballotTypes: SheetOf<BallotType>;
+    }
+  | {
+      type: SheetValidationErrorType.MismatchedElectionHash;
+      electionHashes: SheetOf<ElectionDefinition['electionHash']>;
+    }
+  | {
+      type: SheetValidationErrorType.MismatchedLocales;
+      locales: SheetOf<BallotLocale>;
+    }
+  | {
+      type: SheetValidationErrorType.MismatchedPrecinct;
+      precinctIds: SheetOf<Precinct['id']>;
+    }
+  | {
+      type: SheetValidationErrorType.InvalidPageType;
+      pageTypes: Array<PageInterpretation['type']>;
+    };
+
+/**
+ * Result of a successful validation including the type of ballot and the
+ * relevant interpretations. The interpretations in the `hmpb` type are always
+ * in consecutive, [front, back] order. The `wasReversed` property indicates
+ * whether the sheet passed to the validator was original in reverse order or not.
+ */
+export type ValidatedSheet =
+  | {
+      type: 'bmd';
+      interpretation: InterpretedBmdPage;
+    }
+  | {
+      type: 'hmpb';
+      wasReversed: boolean;
+      interpretation: SheetOf<InterpretedHmpbPage>;
+    };
+
+/**
+ * Validates sheet interpretations
+ */
+export function validateSheetInterpretation([
```

Comment:
Nit: I might call this `normalize` or `canonicalize` since it's actually modifying the input into a standardized form. Or `normalizeAndValidate`

Reply:
skip